onnxruntime/core/framework/session_state_utils.cc

```diff
@@ -92,20 +92,59 @@ common::Status SaveInitializedTensors(
     const GraphViewer& graph, const OrtMemoryInfo& default_cpu_memory_info,
     const OrtValueNameIdxMap& ort_value_name_idx_map, ITensorAllocator& planner,
     const std::function<Status(int idx, const OrtValue& value, const OrtCallback& d, bool constant)>& save_tensor_func,
-    const logging::Logger& logger, const DataTransferManager& data_transfer_mgr) {
+    const logging::Logger& logger, const DataTransferManager& data_transfer_mgr,
+    const ExecutionPlanBase& exec_plan,
+    const SessionOptions& session_options) {
   LOGS(logger, INFO) << "Saving initialized tensors.";
   ORT_ENFORCE(ort_value_name_idx_map.MaxIdx() > -1, "OrtValue indexes should have been populated.");
 
+  // Determine if an intializer was supplied by the user for the purpose of sharing and if it requires a cross-device
+  // copy. In case a cross-device copy is required, sharing cannot be accomplished since we allocate our own buffer
+  // for the destn device which cannot be shared between sessions.
+  auto use_user_supplied_initializer =
+      [&session_options, &exec_plan, &logger, &ort_value_name_idx_map](const std::string& name) -> bool {
+    bool retval = false;
+    auto it = session_options.initializers_to_share_map.find(name);
+    if (it == session_options.initializers_to_share_map.end()) {
+      retval = false;
+    } else {
+      int ort_value_index = -1;
+      if (!ort_value_name_idx_map.GetIdx(name, ort_value_index).IsOK()) {
+        retval = false;
+      } else {
+        auto planned_mem_info = exec_plan.GetLocation(ort_value_index);
+        auto user_mem_info = it->second->Get<Tensor>().Location();
+        retval = user_mem_info.device == planned_mem_info.device;
```

Comment:
1) Can the user still pass in an OrtValue whose memory is on another device (memory owned by the user) as long as the device matches the planned mem info ?

2) What is the ramification of this check ? If ORT decides during planning that the OrtValue needs to be on CUDA and the user supplied an OrtValue whose memory is on CPU - it is just rejected ? Would the user always know where they need to allocate memory for their initializer ?

Reply:
#1: yes.
#2: If the devices are different, we print a warning saying that we won't be able to use the initializer supplied by the user. So, it's rejected in this sense. The model loading, running, etc continues to happen as before. The user can allocate the initializer on a device where he/she wants to run inference on.